File: src/api/facturas.ts
import { callCommand } from './client'
import type {
  Audit,
  CivilDate,
  EstadoInfo,
  ListQuery,
  LookupItem,
  Money,
  PagedResult,
  RowVersion,
  Uuid,
} from './types'

/** See `docs/11-contratos-tauri.md` §5.6. */

export type EstadoFactura =
  'Borrador' | 'Emitida' | 'Pagada' | 'Anulada' | 'Vencida' | 'PagadaParcial'

export interface FacturaFiltro {
  /** Matched against the number and the customer name. */
  texto?: string
  clienteId?: Uuid
  /** Empty means every state; the screen offers a multi-select. */
  estados?: EstadoFactura[]
  fechaDesde?: CivilDate
  fechaHasta?: CivilDate
  soloImpagas?: boolean
  soloVencidas?: boolean
}

/**
 * `total` is sent because the form shows it, but the backend overwrites it with `subtotal + iva`.
 */
export interface FacturaInput {
  numero: string
  fecha: CivilDate
  fechaVencimiento: CivilDate | null
  clienteId: Uuid
  subtotal: Money
  iva: Money
  total: Money
  observaciones: string | null
}

export interface PagoFacturaInput {
  facturaId: Uuid
  fecha: CivilDate
  monto: Money
  medioPago: string
}

export interface PagoFacturaItem {
  id: Uuid
  facturaId: Uuid
  fecha: CivilDate
  monto: Money
  medioPago: string
  rowVersion: RowVersion
}

export interface FacturaListItem {
  id: Uuid
  numero: string
  fecha: CivilDate
  fechaVencimiento: CivilDate | null
  clienteId: Uuid
  clienteNombre: string
  estado: EstadoFactura
  subtotal: Money
  iva: Money
  total: Money
  pagado: Money
  saldo: Money
  diasMora: number
  rowVersion: RowVersion
}

export interface FacturaDetalle {
  id: Uuid
  numero: string
  fecha: CivilDate
  fechaVencimiento: CivilDate | null
  clienteId: Uuid
  clienteNombre: string
  estado: EstadoInfo
  subtotal: Money
  iva: Money
  total: Money
  pagado: Money
  saldo: Money
  diasMora: number
  observaciones: string | null
  pagos: PagoFacturaItem[]
  /** False when the state takes no money, so the payment form is disabled up front. */
  admitePagos: boolean
  puedeEliminarse: boolean
  audit: Audit
}

export const FACTURAS_SORTABLE = [
  'fecha',
  'numero',
  'clienteNombre',
  'estado',
  'total',
  'pagado',
  'saldo',
  'fechaVencimiento',
  'createdAt',
] as const

/** The options the payment form offers. A historical value outside this list is kept as it is. */
export const MEDIOS_PAGO = [
  'Efectivo',
  'Transferencia',
  'Cheque',
  'TarjetaDebito',
  'TarjetaCredito',
  'MercadoPago',
  'Otro',
] as const

export function listFacturas(
  query: ListQuery<FacturaFiltro>,
): Promise<PagedResult<FacturaListItem>> {
  return callCommand('facturas_list', { query })
}

export function getFactura(id: Uuid): Promise<FacturaDetalle> {
  return callCommand('facturas_get', { id })
}

export function createFactura(dto: FacturaInput): Promise<FacturaDetalle> {
  return callCommand('facturas_create', { dto })
}

export function updateFactura(
  id: Uuid,
  dto: FacturaInput,
  rowVersion: RowVersion,
): Promise<FacturaDetalle> {
  return callCommand('facturas_update', { id, dto, rowVersion })
}

export function transitionFactura(
  id: Uuid,
  destino: EstadoFactura,
  rowVersion: RowVersion,
): Promise<FacturaDetalle> {
  return callCommand('facturas_transition', { id, destino, rowVersion })
}

export function deleteFactura(id: Uuid, rowVersion: RowVersion): Promise<void> {
  return callCommand('facturas_delete', { id, rowVersion })
}

export function lookupFacturas(
  clienteId?: Uuid,
  soloImpagas = false,
  texto?: string,
  limite?: number,
): Promise<LookupItem[]> {
  return callCommand('facturas_lookup', { clienteId, soloImpagas, texto, limite })
}

export function listPagos(facturaId: Uuid): Promise<PagoFacturaItem[]> {
  return callCommand('facturas_pagos', { facturaId })
}

// Every payment command answers with the whole invoice: a payment moves the balance and can move
// the state, so returning only the payment would leave the screen showing stale totals.

export function createPago(dto: PagoFacturaInput): Promise<FacturaDetalle> {
  return callCommand('facturas_pago_create', { dto })
}

export function updatePago(
  id: Uuid,
  dto: PagoFacturaInput,
  rowVersion: RowVersion,
): Promise<FacturaDetalle> {
  return callCommand('facturas_pago_update', { id, dto, rowVersion })
}

export function deletePago(id: Uuid, rowVersion: RowVersion): Promise<FacturaDetalle> {
  return callCommand('facturas_pago_delete', { id, rowVersion })
}

File: src-tauri/src/commands/facturas.rs
//! Commands of `facturas` and their payments. See `docs/11-contratos-tauri.md` §5.8.
//!
//! Every payment command answers with the whole invoice: a payment changes the balance and can
//! change the state, and returning only the payment would leave the screen showing stale totals.

use certaro_application::dtos::common::{ListQuery, LookupItem};
use certaro_application::dtos::facturas::{
    FacturaDetalle, FacturaFiltroDto, FacturaInput, FacturaListItem, PagoFacturaInput,
    PagoFacturaItem,
};
use certaro_application::PagedResult;
use certaro_domain::EstadoFactura;
use tauri::State;
use uuid::Uuid;

use crate::error::{handle, ApiResult};
use crate::state::AppState;

#[tauri::command]
pub async fn facturas_list(
    state: State<'_, AppState>,
    query: ListQuery<FacturaFiltroDto>,
) -> ApiResult<PagedResult<FacturaListItem>> {
    let outcome = match state.services() {
        Ok(services) => services.facturas.list(query).await,
        Err(e) => Err(e),
    };
    handle("facturas_list", outcome)
}

#[tauri::command]
pub async fn facturas_get(state: State<'_, AppState>, id: Uuid) -> ApiResult<FacturaDetalle> {
    let outcome = match state.services() {
        Ok(services) => services.facturas.get(id).await,
        Err(e) => Err(e),
    };
    handle("facturas_get", outcome)
}

#[tauri::command]
pub async fn facturas_create(
    state: State<'_, AppState>,
    dto: FacturaInput,
) -> ApiResult<FacturaDetalle> {
    let outcome = match state.services() {
        Ok(services) => services.facturas.create(dto).await,
        Err(e) => Err(e),
    };
    handle("facturas_create", outcome)
}

#[tauri::command]
pub async fn facturas_update(
    state: State<'_, AppState>,
    id: Uuid,
    dto: FacturaInput,
    row_version: String,
) -> ApiResult<FacturaDetalle> {
    let outcome = match state.services() {
        Ok(services) => services.facturas.update(id, dto, &row_version).await,
        Err(e) => Err(e),
    };
    handle("facturas_update", outcome)
}

#[tauri::command]
pub async fn facturas_transition(
    state: State<'_, AppState>,
    id: Uuid,
    destino: EstadoFactura,
    row_version: String,
) -> ApiResult<FacturaDetalle> {
    let outcome = match state.services() {
        Ok(services) => {
            services
                .facturas
                .transition(id, destino, &row_version)
                .await
        }
        Err(e) => Err(e),
    };
    handle("facturas_transition", outcome)
}

#[tauri::command]
pub async fn facturas_delete(
    state: State<'_, AppState>,
    id: Uuid,
    row_version: String,
) -> ApiResult<()> {
    let outcome = match state.services() {
        Ok(services) => services.facturas.delete(id, &row_version).await,
        Err(e) => Err(e),
    };
    handle("facturas_delete", outcome)
}

#[tauri::command]
pub async fn facturas_lookup(
    state: State<'_, AppState>,
    cliente_id: Option<Uuid>,
    solo_impagas: Option<bool>,
    texto: Option<String>,
    limite: Option<u64>,
) -> ApiResult<Vec<LookupItem>> {
    let outcome = match state.services() {
        Ok(services) => {
            services
                .facturas
                .lookup(cliente_id, solo_impagas.unwrap_or(false), texto, limite)
                .await
        }
        Err(e) => Err(e),
    };
    handle("facturas_lookup", outcome)
}

#[tauri::command]
pub async fn facturas_pagos(
    state: State<'_, AppState>,
    factura_id: Uuid,
) -> ApiResult<Vec<PagoFacturaItem>> {
    let outcome = match state.services() {
        Ok(services) => services.facturas.pagos_de(factura_id).await,
        Err(e) => Err(e),
    };
    handle("facturas_pagos", outcome)
}

#[tauri::command]
pub async fn facturas_pago_create(
    state: State<'_, AppState>,
    dto: PagoFacturaInput,
) -> ApiResult<FacturaDetalle> {
    let outcome = match state.services() {
        Ok(services) => services.facturas.crear_pago(dto).await,
        Err(e) => Err(e),
    };
    handle("facturas_pago_create", outcome)
}

#[tauri::command]
pub async fn facturas_pago_update(
    state: State<'_, AppState>,
    id: Uuid,
    dto: PagoFacturaInput,
    row_version: String,
) -> ApiResult<FacturaDetalle> {
    let outcome = match state.services() {
        Ok(services) => {
            services
                .facturas
                .actualizar_pago(id, dto, &row_version)
                .await
        }
        Err(e) => Err(e),
    };
    handle("facturas_pago_update", outcome)
}

#[tauri::command]
pub async fn facturas_pago_delete(
    state: State<'_, AppState>,
    id: Uuid,
    row_version: String,
) -> ApiResult<FacturaDetalle> {
    let outcome = match state.services() {
        Ok(services) => services.facturas.borrar_pago(id, &row_version).await,
        Err(e) => Err(e),
    };
    handle("facturas_pago_delete", outcome)
}
